src/lib.rs

```diff
@@ -61,9 +65,116 @@ impl State {
     }
 }
 
+#[cfg(feature="std")]
+mod oscbuf {
+    use super::MAX_OSC_RAW;
+    pub const DEFAULT_SIZE_LIMIT: usize = u32::max_value() as usize;
+
+    /// When std is available, we can simply use a Vec for the
+    /// OSC storage.  We'll pre-size it to something reasonable.
+    pub struct OscBuffer {
+        buf: Vec<u8>,
+        size_limit: usize,
+    }
+
+    impl std::ops::Deref for OscBuffer {
+        type Target = Vec<u8>;
+        fn deref(&self) -> &Vec<u8> {
+            &self.buf
+        }
+    }
+
+    impl std::ops::DerefMut for OscBuffer {
+        fn deref_mut(&mut self) -> &mut Vec<u8> {
+            &mut self.buf
+        }
+    }
+
+    impl OscBuffer {
+        pub fn new(size_limit: usize) -> Self {
+            Self {
+                buf: Vec::with_capacity(MAX_OSC_RAW.min(size_limit)),
+                size_limit,
+            }
+        }
+
+        pub fn clear(&mut self) {
+            // If we're a bit heavy, shrink our usage.
+            // Ideally we'd call Vec::shrink_to() but that is
+            // nightly only, so we fake it by resizing and then
+            // asking for a shrink to fit the desired capacity.
+            if self.buf.capacity() > 2 * MAX_OSC_RAW {
+                self.buf.resize(MAX_OSC_RAW, 0u8);
+                self.buf.shrink_to_fit();
+            }
+            self.buf.clear();
+        }
+
+        /// Returns true if we are able to store an additional
+        /// byte of data.  Since we have an allocator available,
+        /// we always return true.
+        pub fn have_capacity(&self) -> bool {
+            self.buf.len() < self.size_limit
+        }
+    }
+}
+
+#[cfg(not(feature="std"))]
+mod oscbuf {
+    use super::MAX_OSC_RAW;
+    pub const DEFAULT_SIZE_LIMIT: usize = MAX_OSC_RAW;
+    /// The no_std version of OscBuffer uses a statically allocated
+    /// array and thus has a limit on the size of the data that it
+    /// can hold.  If an OSC exceeds this length, it will be silently
+    /// truncated.
+    /// OscBuffer presents a subset of the Vec API.
+    pub struct OscBuffer {
+        buf: [u8; MAX_OSC_RAW],
+        idx: usize,
+        size_limit: usize,
+    }
+
+    impl OscBuffer {
+        pub fn new(size_limit: usize) -> Self {
+            debug_assert!(size_limit <= MAX_OSC_RAW);
```

Comment:
If we have a limitation like this, it should be a runtime Error/hard panic in all modes or safety should be ensured using the typesystem (like using a u16 instead).

However, I'm not sure if we need to do that here. You said that for a stack implementation you'll run out of stack space anyways for bigger stuff, so why not just stick to the old length of `1024` and hide this complexity from the consumer who will likely have no idea which effects this is going to have?

Reply:
I'll make this an unconditional `assert!`.  Anything else propagates up through the API in a way that I think will be ugly and take more wrangling to resolve than I think is needed.

The size limit needs to be passed down and respected in order for the unit tests to pass in both std and nostd feature modes.

Reply:
> The size limit needs to be passed down and respected in order for the unit tests to pass in both std and nostd feature modes.

I don't think it makes sense to have a worse API just to have tests complete in both modes. I'd say that runtime errors are usually extremely annoying for consumers, so I think it's probably best to stay away from them whenever possible.

Reply:
I understand your perspective, but I'm not clear what it is that you want to be done about it.  Can you give specific advice on how to resolve this?

Reply:
Please also keep in mind that this whole size limit thing is borne out of code review here; it was not in my initial implementation and it is pretty frustrating that you're not in harmony with each other on this point.